cpp/src/structure/sort_and_remove_multi_edges_impl.cuh

```diff
@@ -0,0 +1,317 @@
+struct hash_src_dst_pair {
+  int32_t __device__ operator()(thrust::tuple<vertex_t, vertex_t> t) const
+  {
+    vertex_t pair[2];
+    pair[0] = thrust::get<0>(t);
+    pair[1] = thrust::get<1>(t);
+    cuco::detail::MurmurHash3_32<vertex_t*> hash_func{};
+    return hash_func.compute_hash(reinterpret_cast<std::byte*>(pair), 2 * sizeof(vertex_t));
+  }
+};
+
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+sort_and_remove_multi_edges(raft::handle_t const& handle,
+                            rmm::device_uvector<vertex_t>&& edgelist_srcs,
+                            rmm::device_uvector<vertex_t>&& edgelist_dsts,
+                            size_t mem_frugal_threshold)
+{
+  auto pair_first = thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin());
+
+  if (edgelist_srcs.size() > mem_frugal_threshold) {
+    // Tuning parameter to address high frequency multi-edges
+    size_t num_hash_buckets{16};
+
+    auto hash_counts = compute_hash_sizes(pair_first,
+                                          pair_first + edgelist_srcs.size(),
+                                          num_hash_buckets,
+                                          hash_src_dst_pair<vertex_t>{},
+                                          handle.get_stream());
+
+    auto pivot =
+      static_cast<int32_t>(thrust::distance(hash_counts.begin(),
+                                            thrust::lower_bound(handle.get_thrust_policy(),
+                                                                hash_counts.begin(),
+                                                                hash_counts.end(),
+                                                                edgelist_srcs.size() / 2)));
+
+    auto second_first = detail::mem_frugal_partition(pair_first,
+                                                     pair_first + edgelist_srcs.size(),
+                                                     hash_src_dst_pair<vertex_t>{},
+                                                     pivot,
+                                                     handle.get_stream());
+    thrust::sort(handle.get_thrust_policy(), pair_first, second_first);
+    thrust::sort(handle.get_thrust_policy(), second_first, pair_first + edgelist_srcs.size());
+  } else {
+    thrust::sort(handle.get_thrust_policy(), pair_first, pair_first + edgelist_srcs.size());
+  }
+
+  edgelist_srcs.resize(
+    thrust::distance(pair_first,
+                     thrust::unique(handle.get_thrust_policy(),
+                                    pair_first,
+                                    pair_first + edgelist_srcs.size(),
+                                    [] __device__(auto lhs, auto rhs) {
+                                      return (thrust::get<0>(lhs) == thrust::get<0>(rhs)) &&
+                                             (thrust::get<1>(lhs) == thrust::get<1>(rhs));
+                                    })),
+    handle.get_stream());
```

Comment:
Have you checked thrust::unique does not use a temporary memory buffer (size = original data size). We can test this by creating an array with random data close to 70% of the total memory size and this unique call succeeds.

Reply:
Good catch.  I'll run that test.  I have an idea to work around that.  I'll work on it over the weekend.